cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityManager.java

```diff
@@ -0,0 +1,422 @@
+  private final ProgramLifecycleService programLifecycleService;
+  private final NamespaceAdmin namespaceAdmin;
+  private final MetadataSearchClient metadataSearchClient;
+  private final CConfiguration cConf;
+  private final Lock readLock;
+  private final Lock writeLock;
+
+  @Inject
+  CapabilityManager(CConfiguration cConf, SystemProgramManagementService systemProgramManagementService,
+                    ApplicationLifecycleService applicationLifecycleService, MetadataSearchClient metadataSearchClient,
+                    NamespaceAdmin namespaceAdmin, ProgramLifecycleService programLifecycleService) {
+    this.capabilityStatusMap = new ConcurrentHashMap<>();
+    this.systemProgramManagementService = systemProgramManagementService;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.programLifecycleService = programLifecycleService;
+    this.metadataSearchClient = metadataSearchClient;
+    this.namespaceAdmin = namespaceAdmin;
+    this.cConf = cConf;
+    ReentrantReadWriteLock rwLock = new ReentrantReadWriteLock();
+    readLock = rwLock.readLock();
+    writeLock = rwLock.writeLock();
+  }
+
+  @Override
+  protected void startUp() {
+    LOG.info("Starting {}", this.getClass().getSimpleName());
+    //refresh the status of all capabilities
+    systemProgramManagementService.start();
+    refreshCapabilities();
+  }
+
+  @Override
+  protected void shutDown() {
+    systemProgramManagementService.stopAndWait();
+    LOG.info("Stopping {}", this.getClass().getSimpleName());
+  }
+
+  /**
+   * Return the current status for a capability. If capability is not present, throws {@link IllegalArgumentException}
+   *
+   * @param capability
+   * @return {@link CapabilityStatus}
+   */
+  public CapabilityStatus getStatus(String capability) {
+    if (!capabilityStatusMap.containsKey(capability)) {
+      throw new IllegalArgumentException("Capability not found");
+    }
+    return capabilityStatusMap.get(capability);
+  }
+
+  /**
+   * Returns boolean indicating whether the capability is present in the system
+   *
+   * @param capability
+   * @return boolean indicating presence of capability
+   */
+  public boolean isCapabilityPresent(String capability) {
+    return capabilityStatusMap.containsKey(capability);
+  }
+
+  /**
+   * Applies the given capability configurations
+   *
+   * @param
+   */
+  public void apply(List<CapabilityConfig> capabilityConfigs) {
+    try {
+      writeLock.lock();
+      for (CapabilityConfig capabilityConfig : capabilityConfigs) {
+        String capability = capabilityConfig.getCapability();
+        File file = new File(cConf.get(Constants.Capability.DATA_DIR), capability + ".json");
+        try (FileWriter writer = new FileWriter(file)) {
+          GSON.toJson(capabilityConfig, writer);
+        } catch (IOException ex) {
+          LOG.error("Saving capability {} config to file {} failed with {} ", capability, file, ex);
```

Comment:
I think the exception should be surfaced to the caller, right? Otherwise it's possible changes was detected but never be applied and no one is retrying it.

Reply:
Done.